[mypy] Enable follow_imports = normal for mypy-torch.backends.*

ColinPeppler opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #116311

Summary:

Test Plan:

```
lintrunner --take MYPYINDUCTOR --all-files
ok No lint issues.

lintrunner -a
ok No lint issues.
Successfully applied all patches.
```

Reviewers:

Subscribers:

Tasks:

Tags:

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

ColinPeppler committed:
[mypy] Enable follow_imports = normal for mypy-torch.backends.*

Summary:

Test Plan:

```
lintrunner --take MYPYINDUCTOR --all-files
ok No lint issues.
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

ColinPeppler committed:
Update on "[mypy] Enable follow_imports = normal for mypy-torch.backends.*"


Summary:

Test Plan:

```
lintrunner --take MYPYINDUCTOR --all-files
ok No lint issues.
```

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ColinPeppler committed:
Update on "[mypy] Enable follow_imports = normal for mypy-torch.backends.*"


Summary:

Test Plan:

```
lintrunner --take MYPYINDUCTOR --all-files
ok No lint issues.

lintrunner -a
ok No lint issues.
Successfully applied all patches.
```

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ColinPeppler commented on torch/backends/mkl/__init__.py:
```diff
@@ -45,12 +45,12 @@ def __init__(self, enable):
         if self.enable == VERBOSE_OFF:
             return
-        st = torch._C._verbose.mkl_set_verbose(self.enable)
+        st = torch._C._verbose.mkl_set_verbose(self.enable)  # type: ignore[attr-defined]
```
```
  Error (MYPY) [attr-defined]
    Module has no attribute "_verbose"

        52  |        return self
        53  |
        54  |    def __exit__(self, exc_type, exc_val, exc_tb):
    >>> 55  |        torch._C._verbose.mkl_set_verbose(VERBOSE_OFF)
        56  |        return False
```

I think we could (a) ignore this lint error or (b) tell mypy to pick-up the [_verbose.pyi](https://github.com/pytorch/pytorch/blob/main/torch/_C/_verbose.pyi) stub file.

ColinPeppler replied:
We can do (b) via `import torch._C._verbose as _verbose`

int3 replied:
Yeah we should do b)

ColinPeppler commented on torch/_inductor/fx_passes/mkldnn_fusion.py:
```diff
@@ -1048,7 +1048,10 @@ def forward(self, x):
         """
-        if not (torch.backends.mkldnn.enabled and torch.backends.mkldnn.is_available()):
+        if not (
+            getattr(torch.backends.mkldnn, "enabled", False)
```
```
  Error (MYPYINDUCTOR) [attr-defined]
    Module has no attribute "enabled"

        1048  |
        1049  |        the above's packed weight nodes are duplicate if two linear calls have same input size.
        1050  |        """
    >>> 1051  |        if not (torch.backends.mkldnn.enabled and torch.backends.mkldnn.is_available()):
    ```

int3 replied:
This is due to this bit of code:

https://github.com/pytorch/pytorch/blob/44b98c09ca90ce4529d1a9792b01a0e8afe5345e/torch/backends/mkldnn/__init__.py#L86-L91

In this case, I would resolve it similarly to the dynamo config attributes:

https://github.com/pytorch/pytorch/blob/44b98c09ca90ce4529d1a9792b01a0e8afe5345e/torch/backends/mkldnn/__init__.py#L86-L91

i.e. we can declare `enabled` as a global of `torch.backends.mkldnn` during TYPE_CHECKING. (No need for a separate `.py.in` file here since we only have one attribute)

ColinPeppler committed:
Update on "[mypy] Enable follow_imports = normal for mypy-torch.backends.*"


Summary:

Test Plan:

```
lintrunner --take MYPYINDUCTOR --all-files
ok No lint issues.

lintrunner -a
ok No lint issues.
Successfully applied all patches.
```

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

int3 commented:
hm the review comments still seem unaddressed?

int3 reviewed (changes requested):
putting it back on your queue

ColinPeppler committed:
Update on "[mypy] Enable follow_imports = normal for mypy-torch.backends.*"


Summary:

Test Plan:

```
lintrunner --take MYPYINDUCTOR --all-files
ok No lint issues.

lintrunner -a
ok No lint issues.
Successfully applied all patches.
```

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ColinPeppler commented on torch/backends/mkldnn/__init__.py:
```diff
@@ -86,6 +88,10 @@ def __init__(self, m, name):
 
 
+if TYPE_CHECKING:
+    enabled: ContextProp
```
@int3 re: declare enabled as a global of torch.backends.mkldnn during TYPE_CHECKING

is this what you had in mind?

int3 replied:
yep!

ColinPeppler committed:
Update on "[mypy] Enable follow_imports = normal for mypy-torch.backends.*"


Summary:

Test Plan:

```
lintrunner --take MYPYINDUCTOR --all-files
ok No lint issues.

lintrunner -a
ok No lint issues.
Successfully applied all patches.
```

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ColinPeppler committed:
Update on "[mypy] Enable follow_imports = normal for mypy-torch.backends.*"


Summary:

Test Plan:

```
lintrunner --take MYPYINDUCTOR --all-files
ok No lint issues.

lintrunner -a
ok No lint issues.
Successfully applied all patches.
```

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

int3 commented on torch/backends/mkl/__init__.py:
```diff
@@ -1,4 +1,5 @@
 import torch
+from torch._C import _verbose
```
I thought we were going to do `import torch._C._verbose as _verbose` inside of `./torch/_C/__init__.pyi.in`. I think that's cleaner since it should make both `torch._C._verbose` and `from torch._C import _verbose` typecheck.

ColinPeppler replied:
oh i see, let me try it

ColinPeppler replied:
it works with `from . import _verbose`!

int3 reviewed (approved):
lgtm modulo one nit

ColinPeppler committed:
Update on "[mypy] Enable follow_imports = normal for mypy-torch.backends.*"


Summary:

Test Plan:

```
lintrunner --take MYPYINDUCTOR --all-files
ok No lint issues.

lintrunner -a
ok No lint issues.
Successfully applied all patches.
```

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ColinPeppler commented:
@pytorchbot merge

ColinPeppler commented:
@pytorchbot merge
